pkg/model/project.go

```diff
@@ -243,3 +350,78 @@ func (p *ProjectSSOConfig_GitHub) GenerateAuthCodeURL(project, callbackURL, stat
 
 	return authURL, nil
 }
+
+// FilterRBACRoles filter rbac roles for built-in or not.
+func (p *Project) FilterRBACRoles(isBuiltin bool) []*ProjectRBACRole {
+	v := make([]*ProjectRBACRole, 0, len(p.RbacRoles))
+	for _, role := range p.RbacRoles {
+		if role.IsBuiltin != isBuiltin {
+			continue
+		}
+		v = append(v, role)
+	}
+	return v
+}
+
+// ValidateRBACRoles validate rbac roles.
+func (p *Project) ValidateRBACRoles(roles []*ProjectRBACRole) error {
+	v := make(map[string]struct{}, len(roles))
+	for _, role := range roles {
+		if role.IsBuiltinName() {
+			return fmt.Errorf("cannot use built-in role name")
+		}
+		if _, ok := v[role.Name]; ok {
+			return fmt.Errorf("role name must be unique")
+		}
+		v[role.Name] = struct{}{}
+	}
+	return nil
+}
+
+func (p *ProjectRBACRole) IsBuiltinName() bool {
+	return p.Name == builtinRBACRoleAdmin.String() ||
+		p.Name == builtinRBACRoleEditor.String() ||
+		p.Name == builtinRBACRoleViewer.String()
+}
+
+// ValidateUserGroups validate user groups.
+func (p *Project) ValidateUserGroups(groups []*ProjectUserGroup) error {
+	v := make(map[string]struct{}, len(groups))
+	for _, group := range groups {
+		if _, ok := v[group.SsoGroup]; ok {
+			return fmt.Errorf("cannot assign multi roles to sso group")
+		}
+		v[group.SsoGroup] = struct{}{}
+	}
+	return nil
+}
+
+// SetUserGroup adds a new user group or updates if it already exists.
+func (p *Project) SetUserGroup(group, role string) {
+	v := &ProjectUserGroup{
+		SsoGroup: group,
+		Role:     role,
+	}
+	for i := range p.UserGroups {
+		if p.UserGroups[i].SsoGroup == group {
+			p.UserGroups[i] = v
+			return
+		}
+	}
+	p.UserGroups = append(p.UserGroups, v)
+}
+
+// MigrateFromRBAC migrate rbac config to user group.
+func (p *Project) MigrateFromRBAC() {
+	rbac := p.Rbac
+	if rbac.GetAdmin() != "" {
+		p.SetUserGroup(rbac.Admin, builtinRBACRoleAdmin.String())
+	}
+	if rbac.GetEditor() != "" {
+		p.SetUserGroup(rbac.Editor, builtinRBACRoleEditor.String())
+	}
+	if rbac.GetViewer() != "" {
+		p.SetUserGroup(rbac.Viewer, builtinRBACRoleViewer.String())
+	}
```

Comment:
nit: Use the field directly as our convention.

```suggestion
	if rbac.Admin != "" {
		p.SetUserGroup(rbac.Admin, builtinRBACRoleAdmin.String())
	}
	if rbac.Editor != "" {
		p.SetUserGroup(rbac.Editor, builtinRBACRoleEditor.String())
	}
	if rbac.Viewer != "" {
		p.SetUserGroup(rbac.Viewer, builtinRBACRoleViewer.String())
	}
```